### ISSUE TITLE
String.Substring Unicode

### ISSUE THREAD

**Issue by nu8** [dotnet/dotnet-api-docs#4948]
If you have a file like this:

~~~cs
using System;

class Program {
   static void Main() {
      var s = "📗📕";
      var s1 = s.Substring(0, 2);
      Console.WriteLine(s1 == "📗");
   }
}
~~~

you notice I had to use length 2, as the method counts UTF-16 code units, rather
than characters. I am posting this issue, as I did not see mention of Unicode or
UTF-16 in the documentation for this method.

https://docs.microsoft.com/dotnet/api/system.string.substring

**Comment by carlossanlop** [comment]
Area owners of System.Text.Encoding @layomia @krwq @tarekgh.
cc @GrabYourPitchforks too in case he can also offer some insight.

**Comment by tarekgh** [comment]
String docs is already clear about the contents of the string https://docs.microsoft.com/en-us/dotnet/api/system.string?view=netcore-3.1#remarks. I am not sure if we have to repeat that in every API inside String class.

**Comment by nu8** [comment]
To argue against myself, I would say that the [JavaScript substring][1] doesnt
mention Unicode either.

But [JavaScript length][2] does. As well as the JavaScript [Text formatting
guide][3], similar to the link @tarekgh put.

I dont think @tarekgh argument holds water. Users should know what behavior to
expect from a method, without resulting to trial and error, and without having
to visit other pages to do so. As a matter of personal opinion, I dont think
[SMP Unicode][4] characters should be treated as second class citizen, as they
are with C#. I understand that some of that is unavoidable, as it was a design
decision of C# to use UTF-16 strings.

But I think despite this shortcoming, it can at least be documented, so that
people know up front what is going to happen, instead of after the fact.

[1]://developer.mozilla.org/Web/JavaScript/Reference/Global_Objects/String/substring
[2]://developer.mozilla.org/Web/JavaScript/Reference/Global_Objects/String/length
[3]://developer.mozilla.org/Web/JavaScript/Guide/Text_formatting
[4]://wikipedia.org/wiki/Plane_(Unicode)

**Comment by tarekgh** [comment]
My worries here is users understand more `char` than 'code unit'. I know they are different (from Unicode definition) but they are same as System.Char is a code unit. Not every API user understand the difference. 

I don't mind adding adding extra sentence in the remark section clarifying the behavior but we have to pick the language that don't confuse users who are not familiar with the definition.

@nu8 feel free to open a doc PR with what you suggest and we can go from there.

**Comment by GrabYourPitchforks** [comment]
I think the doc https://docs.microsoft.com/dotnet/standard/base-types/character-encoding-introduction explains this concept fairly well, including the fact that `char` is a _UTF-16 code unit_. This documentation is also linked to from the _Remarks_ section of the `char` documentation (https://docs.microsoft.com/dotnet/api/system.char). Maybe it'd also be good to link to it from the _Remarks_ section of the `string` documentation (https://docs.microsoft.com/dotnet/api/system.string)?

**Comment by nu8** [comment]
If this:

https://docs.microsoft.com/dotnet/standard/base-types/character-encoding-introduction

was linked here:

https://docs.microsoft.com/en-us/dotnet/api/system.string.substring

under See Also or similar, that would be a good compromise. I can understand not wanting the same thing copied everywhere, but I think it important that its mentioned or linked *somewhere* on the page.

**Comment by GrabYourPitchforks** [comment]
I don't think it belongs on the `Substring` doc. String length is given in terms of chars (code units), strings are indexed by chars (code units), and so on. This discussion really belongs on the `string` class or the `char` class - or both.

Once the documentation makes clear that a `string` is really just an immutable sequence of `char`s, then it naturally follows that any operation that deals with indices (`Substring`, `IndexOf`, `Remove`, `CopyTo`, etc.) also operates on chars (code units) rather than scalar values. There's no need to annotate every method on `string` to convey this concept.

**Comment by nu8** [comment]
I do notice that for this page:

https://docs.microsoft.com/dotnet/api/system.string

under the heading Returns, you have this:

> [String](//docs.microsoft.com/dotnet/api/system.string)
>
> A string that is equivalent to the substring that begins at startIndex in this
> instance, or Empty if startIndex is equal to the length of this instance.

Where if you follow the link, it does say:

> Represents text as a sequence of UTF-16 code units.

I think it would be pretty low impact to change to something like this:

> [String](//docs.microsoft.com/dotnet/api/system.string)
>
> A string (sequence of UTF-16 code units) that is equivalent to the substring
> that begins at startIndex in this instance, or Empty if startIndex is equal to
> the length of this instance.
